internal/provider/resource_gitlab_repository_file.go

```diff
@@ -89,10 +117,16 @@ var _ = registerResource("gitlab_repository_file", func() *schema.Resource {
 })
 
 func resourceGitlabRepositoryFileCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*gitlab.Client)
 	project := d.Get("project").(string)
 	filePath := d.Get("file_path").(string)
 
+	log.Printf("[DEBUG] gitlab_repository_file: waiting for lock to create %s/%s", project, filePath)
+	apiSync.Lock()
```

Comment:
One limitation of using `sync.Mutex` here is that it will block "forever". Since the caller passes a `context.Context`, it expects that you'll return when the context is cancelled.

One idea would be to use a channel as a lock.

```go

type lock chan struct{}

func newLock() lock {
	// The buffer size of one allows us to use it as a lock.
	return make(lock, 1)
}

func (c lock) lock(ctx context.Context) error {
	select {
	case c <- struct{}{}:
		// Lock acquired.
		return nil
	case <-ctx.Done():
		// Timeout.
		return ctx.Err()
	}
}

func (c lock) unlock() {
	<-c
}

var resourceGitlabRepositoryFileLock = newLock()

func resourceGitlabRepositoryFileCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
	if err := resourceGitlabRepositoryFileLock.lock(ctx); err != nil {
		return diag.FromErr(err)
	}
	defer resourceGitlabRepositoryFileLock.unlock()
}
```

Reply:
I actually was thinking of something like this ...

> One limitation of using sync.Mutex here is that it will block "forever"

... however, in which case could that possibly happen for us? Given that the GitLab calls (which may block) are using the passed ctx?

I can implement it as kind of a "defensive" thing if you think it makes sense.

Reply:
You are correct that there cannot be deadlocks. By ""forever"" I mean that it will block for longer than the caller might expect, because we are not respecting the context.

As an exaggerated example, suppose there are two `gitlab_repository_file` resources with different timeouts.

```terraform
resource "gitlab_repository_file" "foo" {
  timeouts {
    create = "60m"
  }
}

resource "gitlab_repository_file" "bar" {
  timeouts {
    create = "5m"
  }
}
```

If `"foo"` gets the lock first and blocks, then `"bar"` will also timeout after `60m` instead of the intended `5m`.

I would be ok with keeping the `sync.Mutex` if it as a temporary hack. Otherwise I prefer the channel lock.

For me, generally I only use `sync.Mutex` to protect data structures like maps when I am not doing I/O while holding the lock.

Reply:
Ah, yes makes sense. I'll adapt it tomorrow :)

Reply:
Alright, looks good now :)